cidc_api/utils/migrations.py

```diff
@@ -0,0 +1,159 @@
+            try:
+                task.do()
+            except:
+                for done_task in self.tasks[:i]:
+                    done_task.undo()
+                raise
+
+
+def run_metadata_migration(metadata_migration: Callable[[dict], MigrationResult]):
+    """Migrate trial metadata, upload job patches, and downloadable files according to `metadata_migration`"""
+    session = Session(bind=op.get_bind())
+
+    # Initialize an empty list of pieces of work to do on GCS resources
+    gcs_tasks = RollbackableQueue()
+
+    # Migrate all trial records
+    trials: List[TrialMetadata] = session.query(TrialMetadata).with_for_update().all()
+    for trial in trials:
+        migration = metadata_migration(trial.metadata_json)
+
+        # Update the trial metadata object
+        trial.metadata_json = migration.result
+
+        # Update the relevant downloadable files and GCS objects
+        for old_gcs_uri, artifact in migration.file_updates.items():
+            # Update the downloadable file associated with this blob
+            df: DownloadableFiles = (
+                session.query(DownloadableFiles)
+                .filter_by(object_url=old_gcs_uri)
+                .with_for_update()
+                .one()
+            )
+            for column, value in artifact.items():
+                if hasattr(df, column):
+                    setattr(df, column, value)
+
+            # Regenerate additional metadata from the migrated clinical trial
+            # metadata object.
+            df.additional_metadata = _get_uuid_info(
+                migration.result, artifact["upload_placeholder"]
+            )[1]
+
+            # If the GCS URI has changed, rename the blob
+            new_gcs_uri = artifact["object_url"]
+            if old_gcs_uri != new_gcs_uri:
+                renamer = PieceOfWork(
+                    lambda: rename_gcs_blob(
+                        GOOGLE_DATA_BUCKET, old_gcs_uri, new_gcs_uri
+                    ),
+                    lambda: rename_gcs_blob(
+                        GOOGLE_DATA_BUCKET, new_gcs_uri, old_gcs_uri
+                    ),
+                )
+                gcs_tasks.schedule(renamer)
+
+    # Migrate all assay upload successes
+    successful_assay_uploads: List[AssayUploads] = session.query(
+        AssayUploads
+    ).filter_by(status=AssayUploadStatus.MERGE_COMPLETED.value).with_for_update().all()
+    for upload in successful_assay_uploads:
+        migration = metadata_migration(upload.assay_patch)
+
+        # Update the metadata patch
+        upload.assay_patch = migration.result
+
+        # Update the GCS URIs of files that were part of this upload
+        old_file_map = upload.gcs_file_map
+        new_file_map = {}
+        for (
+            old_upload_uri,
+            old_target_uri,
+            artifact_uuid,
+        ) in upload.upload_uris_with_data_uris_with_uuids():
+            upload_timestamp = old_upload_uri[len(old_target_uri) + 1 :]
```

Comment:
nit: I would abstract that into a helper function. at some point I wanted to add extensions to those upload filenames as say `uploads/staging_0/olink/chip_1chip/assay_raw_ct.xlsx/2019-10-07T20:00:40.903770.xlsx` so they will be opened right away

Reply:
I'd rather do this in a separate PR, since this helper function (and an associated helper for building URLs) should be used everywhere functionality like this exists in the API code.